### DISCUSSION TITLE
Add pnpm support to quasar-cli?

### DISCUSSION THREAD

**Discussion by sullyD64** [quasarframework/quasar#11059] [Ideas / Proposals] [answered]
Hi everyone. First of all, thank you for releasing and maintaining such an awesome gem in the vue ecosystem. 

At my company we're trying to build several quasar apps which will share components and scripts also built on top of quasar. We are trying the monorepo approach with rushjs + `pnpm`. Albeit we know quasar recommends yarn (and seemingly doesn't support pnpm), we value pnpm's for its performance and strictness over dependency management (also because it works very well in monorepos).  

However, we are encountering lots of errors (missing package "x")  during both `quasar dev` and `quasar build` . We've read this is caused by [phantom dependencies](https://rushjs.io/pages/advanced/phantom_deps/), which `@quasar/app` seems to have in large numbers (not counting webpack-dev-server). We've tried the several fixes listed in the pnpm docs, but we have some hard time fixing individual phantom dependencies without recurring to the `-shamefully-hoist` option, which basically defeats the purpose of being strict about dependencies anyway. 

With that said, we are wondering if there are any plans for quasar-cli to support pnpm, which will require the cli to explicitly add the dependencies which are internally used when building or serving the app (for comparison, `vue-cli` adds `vue` to `dependencies` and `@vue/compiler-sfc` to `devDependencies`, and so on..). 

P.s: we could try to go with the vue-cli quasar plugin, but we value quasar-cli's capability as much as pnpm!

**Comment by yusufkandemir** [comment, marked as answer]
Update: pnpm and bun package managers are supported.

pnpm:
- [@quasar/cli-v1.4.0](https://github.com/quasarframework/quasar/releases/tag/%40quasar%2Fcli-v1.4.0)
- [@quasar/app-vite-v1.1.0](https://github.com/quasarframework/quasar/releases/tag/%40quasar%2Fapp-vite-v1.1.0)
- [@quasar/app-webpack-v3.6.0](https://github.com/quasarframework/quasar/releases/tag/%40quasar%2Fapp-webpack-v3.6.0)

bun:
- [@quasar/cli-v2.3.0](https://github.com/quasarframework/quasar/releases/tag/%40quasar%2Fcli-v2.3.0)
- [@quasar/app-vite-v1.6.0](https://github.com/quasarframework/quasar/releases/tag/%40quasar%2Fapp-vite-v1.6.0)
- [@quasar/app-webpack-v3.11.0](https://github.com/quasarframework/quasar/releases/tag/%40quasar%2Fapp-webpack-v3.11.0)


Old answer:
I believe it's similar to Yarn v2 Plug'N'Play in terms of how dependencies are handled. Please read this for a thorough explanation:
https://github.com/quasarframework/quasar/pull/10277#issuecomment-942534889

**Reply by sullyD64** [reply]
Hi @yusufkandemir and thanks for answering. I've read through #10277 and i totally understand their point, which is that if they test the dependencies themselves and pin them, then users of q/app can rely without hassle on the transitive deps that quasar uses internally.

Still I think supporting pnpm (albeit forcing `-shamefully-hoist`) as an option for installing the generated deps would be a fine addition to the cli. This is similar to what `vue-cli` does when using `--packageManager pnpm`.

I look forward to pnpm / yarn v2 support for the upcoming Vite-based quasar-cli. For now we'll `-shamefully-hoist` minus the shame :smile: .

**Comment by taai** [comment]
I tried using Quasar with *pnpm* version 7.0.1. It works. :+1:  Sort of. :man_shrugging: 

If you want to use **globally installed** *Quasar CLI* with *pnpm*, I have 3 scenarios for you.

### Scenario 1

You want to use *pnpm* without using ```shamefully-hoist``` setting (as you should).

While reading [Quasar documentation](https://quasar.dev/start/quasar-cli#running-without-the-global-quasar-cli), I spotted that (in case of using *Yarn*) you can prefix the command with ```yarn```, so I tried to prefix the command with ```pnpm``` and it worked.

```bash
# install dependencies
pnpm install

# update dependencies
pnpm update

# development
pnpm quasar dev

# build
pnpm quasar build
```

### Scenario 2

You want or **need** to use *pnpm* **with** using ```shamefully-hoist``` setting, but you don't want to add ```--shamefully-hoist``` flag every time you install or update dependencies.

You have to create a file ```.npmrc``` in the root directory of your project and write this line in that file:
```
shamefully-hoist=true
```

In this scenario you don't have to (but you can) prefix the commands with ```pnpm```.

```bash
# install dependencies
pnpm install

# update dependencies
pnpm update

# development
quasar dev
# or
# pnpm quasar dev

# build
quasar build
# or
# pnpm quasar build
```

### Scenario 3

You want to use *pnpm* and run the commands normally without doing anything specific.

Wait until someone fixes the issue by doing what I [described in a comment](https://github.com/quasarframework/quasar/issues/7581#issuecomment-1123756953). :wink:

**Reply by marty0678** [reply]
~~Thanks for this! It seems Option 1 doesn't work anymore with Quasar 2.14.2 and pnpm 8.14.0. Running `pnpm quasar build` will fail to import any module that is in the nested .pnpm folder.~~

~~Option 2 does work though.~~

Nvm, this was not related to Quasar, it was related to a specific sub dep being used elsewhere directly and not hoisted up with `public-hoist-pattern[]=*package-name*`.

**Reply by TobyMosque** [reply]
pnpm and bun will be options in the next CLI/app.

**Reply by yusufkandemir** [reply]
@TobyMosque pnpm and bun are already supported in the current global and local CLIs.

Newly created projects have it scaffolded, but for old projects, create `.npmrc` with the following contents:
```npmrc
shamefully-hoist=true
```

**Reply by marty0678** [reply]
Yeah I actually just set up a new quasar app a few moments ago and pnpm worked just fine and created this npmrc file:

```
# pnpm-related options
shamefully-hoist=true
strict-peer-dependencies=false
```

Which on that point, is pnpn and bun still technically experimental for Quasar? The docs still say they are but they seem fully supported now.

**Reply by yusufkandemir** [reply]
I don't consider at least the pnpm support as experimental because I have been using it on a lot of projects for a long while. But, it's been supported for less time than npm/yarn. We will probably mention them as stable in the docs after releasing the next major CLIs.
